### ISSUE TITLE
ComposedLinearOperator with DiagonalLinearOperator instantiates n x n matrix.

### ISSUE THREAD

**Issue by quattro** [google/lineax#215]
Hi, I've been working on [glmax](https://github.com/mancusolab/glmax) and during benchmarking I noticed runtimes were slower than expected. After some digging through the HLO, I found the issue to be that the DiagonalLinearOperator was having  `as_matrix` called from the ComposedLinearOperator `as_matrix`. 

This could be addressed in principle with some edge case handling in the ComposedLinearOperator, but I'm not sure how much optimization you'd like to exist there. Given that, this is a fairly common use case (diagonal weighting of observations for IRLS or Newton optimization for GLMs). 

Minimum working example: DiagonalLinearOperator materializes as n×n inside while_loop.

Expected behavior
-----------------
``lx.DiagonalLinearOperator(d) @ lx.MatrixLinearOperator(X)`` represents
``diag(d) @ X``. Applied to a vector it should cost O(np) via elementwise
scaling, never forming an n×n intermediate. When ``lx.Normal(lx.Cholesky())``
wraps this operator the normal equations are

    A^T A = (diag(d) X)^T (diag(d) X) = X^T diag(d^2) X

which costs O(np^2) to form — still no n×n intermediate required.

Observed behavior (reported for lineax)
-----------------------------------------
Inside ``jax.lax.while_loop``, XLA materializes ``lx.DiagonalLinearOperator``
as a full n×n dense matrix, turning the per-iteration cost from O(np^2) into
O(n^2 p). At n=500, p=10 the theoretical ratio is n/p = 50×, matching the
10–20× real-world slowdown observed when benchmarking glmax against statsmodels.

HLO evidence (n=500, p=10, f64)
---------------------------------
Pre-loop:
  ``%iota_compare_fusion = pred[500,500]``  ← identity mask carried as state

Per-iteration body:
  ``%broadcast_select_fusion.1 = f64[500,500]``  ← full diagonal matrix formed
  ``%dot_general.14 = f64[500,10] dot(%broadcast_select_fusion.1, X)``
                              ↑ (n×n)@(n×p)  instead of elementwise scale

```python
import re

import jax
import jax.numpy as jnp
import lineax as lx
import numpy as np


# Enable 64-bit precision (required to match the glmax production context).
jax.config.update("jax_enable_x64", True)


# ── problem dimensions ────────────────────────────────────────────────────────
N = 500   # number of observations
P = 10    # number of predictors


def _make_data(n: int = N, p: int = P, seed: int = 0):
    rng = np.random.default_rng(seed)
    X = rng.standard_normal((n, p))
    beta_true = rng.standard_normal(p) * 0.3
    w = np.exp(X @ beta_true)   # positive weights, like Poisson GLM weights
    z = rng.standard_normal(n)  # adjusted response
    return (
        jnp.array(X, dtype=jnp.float64),
        jnp.array(w, dtype=jnp.float64),
        jnp.array(z, dtype=jnp.float64),
    )


# ── minimal while_loop that mirrors IRLS ─────────────────────────────────────

@jax.jit
def irls_loop_diag(X: jax.Array, w0: jax.Array, z: jax.Array) -> jax.Array:
    """Loop using DiagonalLinearOperator @ MatrixLinearOperator (current glmax)."""
    Xop = lx.MatrixLinearOperator(X)
    solver = lx.Normal(lx.Cholesky())

    def body_fn(carry):
        w, i = carry
        Woh = lx.DiagonalLinearOperator(jnp.sqrt(w))
        A = Woh @ Xop
        b = Woh.mv(z)
        beta = lx.linear_solve(A, b, solver=solver).value
        w_new = jnp.abs(X @ beta) + 1e-3  # simulate weight update
        return (w_new, i + 1)

    def cond_fn(carry):
        _, i = carry
        return i < 5

    w_final, _ = jax.lax.while_loop(cond_fn, body_fn, (w0, 0))
    return w_final


@jax.jit
def irls_loop_direct(X: jax.Array, w0: jax.Array, z: jax.Array) -> jax.Array:
    """Loop forming normal equations directly (no DiagonalLinearOperator)."""
    solver = lx.Cholesky()

    def body_fn(carry):
        w, i = carry
        WX  = w[:, None] * X          # (n, p)
        H   = X.T @ WX                # (p, p)
        rhs = X.T @ (w * z)           # (p,)
        beta = lx.linear_solve(
            lx.MatrixLinearOperator(H, lx.positive_semidefinite_tag),
            rhs, solver=solver).value
        w_new = jnp.abs(X @ beta) + 1e-3
        return (w_new, i + 1)

    def cond_fn(carry):
        _, i = carry
        return i < 5

    w_final, _ = jax.lax.while_loop(cond_fn, body_fn, (w0, 0))
    return w_final


# ── helpers ───────────────────────────────────────────────────────────────────

def _largest_intermediate(hlo: str) -> int:
    """Return the largest dimension seen in any array shape in the HLO."""
    dims = [int(d) for d in re.findall(r"f(?:32|64)\[(\d+)(?:,\d+)*\]", hlo)]
    return max(dims, default=0)


def _count_nxn(hlo: str, n: int) -> int:
    return len(re.findall(rf"f(?:32|64)\[{n},{n}\]", hlo))


def _dot_shapes(hlo: str) -> list[str]:
    lines = []
    for m in re.finditer(r"(f(?:32|64)\[\d+(?:,\d+)*\]\s+dot\([^)]+\))", hlo):
        lines.append(m.group(1).strip())
    return lines


def _cost_analysis(compiled) -> str:
    try:
        costs = compiled.cost_analysis()
        if isinstance(costs, list):
            flops = sum(c.get("flops", 0) for c in costs)
            mem   = sum(c.get("bytes accessed", 0) for c in costs)
        elif isinstance(costs, dict):
            flops = costs.get("flops", 0)
            mem   = costs.get("bytes accessed", 0)
        else:
            return "  (unrecognised cost_analysis format)"
        return f"  FLOPs: {flops:,.0f}    bytes accessed: {mem:,.0f}"
    except Exception as exc:
        return f"  (cost_analysis unavailable: {exc})"


# ── main ──────────────────────────────────────────────────────────────────────

def main():
    X, w0, z = _make_data()

    print(f"Problem: n={N}, p={P}")
    print(f"Theoretical cost ratio (n/p): {N//P}×\n")

    for label, fn in [
        ("DiagonalLinearOperator (current)", irls_loop_diag),
        ("Direct normal equations (proposed fix)", irls_loop_direct),
    ]:
        compiled = fn.lower(X, w0, z).compile()
        hlo = compiled.as_text() or ""

        nxn     = _count_nxn(hlo, N)
        max_dim = _largest_intermediate(hlo)
        dots    = _dot_shapes(hlo)

        print(f"── {label} ──")
        print(_cost_analysis(compiled))
        print(f"  n×n ({N}×{N}) arrays in HLO : {nxn}  {'← BUG' if nxn else '✓ none'}")
        print(f"  Largest intermediate dim   : {max_dim}")
        if dots:
            print("  dot_general shapes found:")
            for d in dots[:6]:
                print(f"    {d}")
        print()

    print("=== Summary ===")
    print(f"  If the DiagonalLinearOperator variant shows f64[{N},{N}] in its HLO,")
    print("  the diagonal operator is being materialized as a full matrix.")
    print("  The direct normal-equations variant should show only f64[P,P] intermediates.")


if __name__ == "__main__":
    main()
```

**Comment by jpbrodrick89** [comment]
Have you tried the GitHub main version of lineax there's a chance we fixed this recently (we've touched ComposedLinearOperator and Cholesky recently)? Also let me know if my job/diagfromjac branch changes anything. If not I'm happy to take a closer look.

**Comment by quattro** [comment]
Hi @jpbrodrick89 I hadn't! I was using the latest release, but installing from github and re-running the above produces: 
```
── DiagonalLinearOperator (current) ──
  FLOPs: 134,346    bytes accessed: 301,965
  n×n (500×500) arrays in HLO : 0  ✓ none
  Largest intermediate dim   : 500

── Direct normal equations (proposed fix) ──
  FLOPs: 126,864    bytes accessed: 285,012
  n×n (500×500) arrays in HLO : 0  ✓ none
  Largest intermediate dim   : 500
```

I can close this as resolved! Thank you.
